### PR TITLE
Clarify cryptic bitwise operations in Partiall-determined haplotype EventGroup subclass

### PR DESCRIPTION
**Pull request by davidbenjamin**
@jamesemery Another DRAGEN PR for you.  No change in output, but it lays some of the groundwork for having multiple determined events at one locus and will make addition of more features easier.

It runs just as fast as before -- the bit-twiddling is encapsulated, but it's still there.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import org.broadinstitute.hellbender.tools.walkers.genotyper.GenotypeAlleleCounts;
+import org.broadinstitute.hellbender.tools.walkers.genotyper.GenotypeIndexCalculator;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * This class is very much like the standard library BitSet class in java.util, but instead of using
+ * a long[] array it uses single (32-bit) ints, which are much faster.  This limits the capacity but
+ * for small sets it makes sense.
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java:12]: Make the comment about this louder and clearer.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 2]
```diff
@@ -806,86 +788,60 @@ public void populateBitset(List<List<Event>> disallowedEvents) {
...
 
-            List<Integer> ints = new ArrayList<>();
+            final List<SmallBitSet> allowedAndDetermined = new ArrayList<>();
             // Iterate from the BACK of the list (i.e. ~supersets -> subsets)
             // NOTE: we skip over 0 here since that corresponds to ref-only events, handle those externally to this code
-            outerLoop:
-            for (int i = allowedEvents.length(); i > 0; i--) {
-                // If the event is allowed AND if we are looking for a particular event to be present or absent.
-                if (allowedEvents.get(i) && (eventMask == 0 || ((i & eventMask) == maskValues))) {
+            for (final SmallBitSet subset = SmallBitSet.fullSet(eventsInOrder.size()); !subset.isEmpty(); subset.decrement()) {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:803]: re-reading this eventhough I wrote it using i-- as a set it occurs to me just how confusing this behavior is... make it a bit clearer in the comments here that this makes sense.

**Review by jamesemery — APPROVED** [review]
👍 A few documentation requests and exactly one material comment that if the bitset object is going to be limited to 30 elements you should put more guardrails into the thing (unless you want to make it a private class) to protect against having to find out what an overflow does here...

File: src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+    public SmallBitSet(final Collection<Integer> elements) {
+        bits = 0;
+        for (final int element : elements) {
+            bits |= elementIndex(element);
+        }
+    }
+
+    // create a full bit set of all 1s in binary up to a certain number of elements i.e. 00000000000111111....
+    public static SmallBitSet fullSet(final int numElements) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java:60]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 4]
```diff
@@ -139,6 +136,7 @@ public static AssemblyResultSet generatePDHaplotypes(final AssemblyResultSet sou
              */
             for (int determinedAlleleIndex = (pileupArgs.determinePDHaps?0:-1); determinedAlleleIndex < allEventsHere.size(); determinedAlleleIndex++) { //note -1 for I here corresponds to the reference allele at this site
                 final boolean isRef = determinedAlleleIndex == -1;
+                final Set<Event> determinedEvents = isRef ? Set.of() : Set.of(allEventsHere.get(determinedAlleleIndex));
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:139]: Yes, this is just the beginning. . .

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 5]
```diff
@@ -806,86 +788,60 @@ public void populateBitset(List<List<Event>> disallowedEvents) {
...
-            if(eventMask==0) {
-                cachedEventLists = Collections.unmodifiableList(output);
+            if(locusOverlapSet.isEmpty()) {
+                cachedEventSets = Collections.unmodifiableList(output);
             }
             return output;
         }
 
+        // create the SmallBitSet of those elements from some collection of events that overlap this EventGroup
+        private SmallBitSet overlapSet(final Collection<Event> events) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:831]: done

File: src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java:12]: done

File: src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        final SmallBitSet result = new SmallBitSet();
+        result.bits = (1 << numElements) - 1;
+        return result;
+    }
+
+    public SmallBitSet increment() {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/SmallBitSet.java:66]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 2 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:803]: done

**Pull request merged**